staging/src/k8s.io/apiserver/pkg/storage/cacher/time_budget.go

```diff
@@ -48,43 +50,47 @@ type timeBudgetImpl struct {
 	sync.Mutex
 	budget time.Duration
 
-	refresh   time.Duration
-	maxBudget time.Duration
+	refresh         time.Duration
+	maxBudget       time.Duration
+	clock           clock.Clock
+	lastRefreshTime time.Time
 }
 
-func newTimeBudget(stopCh <-chan struct{}) timeBudget {
-	result := &timeBudgetImpl{
-		budget:    time.Duration(0),
-		refresh:   refreshPerSecond,
-		maxBudget: maxBudget,
-	}
-	go result.periodicallyRefresh(stopCh)
-	return result
-}
-
-func (t *timeBudgetImpl) periodicallyRefresh(stopCh <-chan struct{}) {
-	ticker := time.NewTicker(time.Second)
-	defer ticker.Stop()
-	for {
-		select {
-		case <-ticker.C:
-			t.Lock()
-			if t.budget = t.budget + t.refresh; t.budget > t.maxBudget {
-				t.budget = t.maxBudget
-			}
-			t.Unlock()
-		case <-stopCh:
-			return
-		}
+func newTimeBudget(clk clock.Clock) timeBudget {
+	return &timeBudgetImpl{
+		budget:          time.Duration(0),
+		refresh:         refreshPerSecond,
+		maxBudget:       maxBudget,
+		clock:           clk,
+		lastRefreshTime: clk.Now(), // set the last refresh time to now
 	}
 }
 
 func (t *timeBudgetImpl) takeAvailable() time.Duration {
 	t.Lock()
 	defer t.Unlock()
-	result := t.budget
+
+	// freeze the time of refresh
+	now := t.clock.Now()
+
+	// number of elapsed seconds since last refresh
+	// it is rounded down to the previous integral second
+	elapsedSeconds := int(now.Sub(t.lastRefreshTime).Seconds())
```

Comment:
That doesn't work - if we are calling takeAvailable more often than once per second, this will always be zero and we will never refresh anything...

Reply:
Yes. You're right. Periodic refreshes within a second of interval shouldn't reset the accumulated budget and `lastRefreshTime`.

Reply:
So, here is what I was thinking:

```
if `elapsedSeconds` is less than 1,
     don't reset the last refresh time
else
     reset the last refresh time to now
```

The accumulated budget will be returned in both cases.

Reply:
I added a test scenario for this specific case as well.

Reply:
It still doesn't work.

If I will be calling that every 1.99s, I will always be adding <time budget>.
So effectively, during 199s, I will add 101*<time budget> instead of 199*<time budget>

This can be mitigated, but I would really like this to be done in more continuous way, given we already touch this code.